src/runtime/simulator.cc

```diff
@@ -39,29 +39,841 @@ bool ParallelConfig::is_data_parallel() const
+      Comm_device *membus = new Comm_device(membus_name, Comm_device::MEMBUS_COMM, node_id, socket_id, device_id, latency, bandwidth);
+      membuses.push_back(membus);
+    }
+  }
+}
+
+void MachineModel_new::add_upis(float latency, float bandwidth)
+{
+  for (int i = 0; i < num_nodes; i++) {
+    int node_id = i;
+    for (int j = 0; j < num_sockets_per_node; j++) {
+      int socket_id = i * num_sockets_per_node + j;
+      int device_id = socket_id;
+      std::string upi_in_name = "UPI_IN " + std::to_string(device_id);
+      Comm_device *upi_in = new Comm_device(upi_in_name, Comm_device::UPI_IN_COMM, node_id, socket_id, device_id, latency, bandwidth);
+      upi_ins.push_back(upi_in);
+      std::string upi_out_name = "UPI_OUT " + std::to_string(device_id);
+      Comm_device *upi_out = new Comm_device(upi_out_name, Comm_device::UPI_OUT_COMM, node_id, socket_id, device_id, latency, bandwidth);
+      upi_outs.push_back(upi_out);
+    }
+  }
+}
+
+void MachineModel_new::add_nics(float latency, float bandwidth, NicDistribution nic_distribution)
+{
+  if (nic_distribution == PER_NODE) {
+    for (int i = 0; i < num_nodes; i++) {
+      int node_id = i;
+      for (int j = 0; j < num_sockets_per_node; j++) {
+        int socket_id = i * num_sockets_per_node + j;
+        int device_id = socket_id;
+        Comm_device *nic_in;
+        Comm_device *nic_out;
+        if (j == 0) {
+          std::string nic_in_name = "NIC_IN " + std::to_string(device_id);
+          nic_in = new Comm_device(nic_in_name, Comm_device::NIC_IN_COMM, node_id, socket_id, device_id, latency, bandwidth);
+          nic_ins.push_back(nic_in);
+          std::string nic_out_name = "NIC_OUT " + std::to_string(device_id);
+          nic_out = new Comm_device(nic_out_name, Comm_device::NIC_OUT_COMM, node_id, socket_id, device_id, latency, bandwidth);
+          nic_outs.push_back(nic_out);
+        }
+        else {
+          nic_ins.push_back(nic_in);
+          nic_outs.push_back(nic_out);
+        }
+      }
+    }
+  }
+  else if (nic_distribution == PER_SOCKET) {
+    for (int i = 0; i < num_nodes; i++) {
+      int node_id = i;
+      for (int j = 0; j < num_sockets_per_node; j++) {
+        int socket_id = i * num_sockets_per_node + j;
+        int device_id = socket_id;
+        std::string nic_in_name = "NIC_IN " + std::to_string(device_id);
+        Comm_device *nic_in = new Comm_device(nic_in_name, Comm_device::NIC_IN_COMM, node_id, socket_id, device_id, latency, bandwidth);
+        nic_ins.push_back(nic_in);
+        std::string nic_out_name = "NIC_OUT " + std::to_string(device_id);
+        Comm_device *nic_out = new Comm_device(nic_out_name, Comm_device::NIC_OUT_COMM, node_id, socket_id, device_id, latency, bandwidth);
+        nic_outs.push_back(nic_out);
+      }
+    }
+  }
+  else {
+    assert(false && "Unknown nic distribution type");
+  }
+}
+
+void MachineModel_new::add_pcis(float latency, float bandwidth)
+{
+  for (int i = 0; i < num_nodes; i++) {
```

Comment:
```suggestion
  for (int node_id = 0; node_id < num_nodes; node_id++) {
```

Reply:
In general if we can substitute out most of the explicit indices with real names the code would be much easier to read